File: prj3_api/src/models/course.model.ts
import mongoose from "mongoose";

const schema = new mongoose.Schema({
    courseId: String,
    name: String,
    midFactor: Number,
    mid: Number,
    final: Number,
    courseFactor: Number,
    student: { type: mongoose.Schema.Types.ObjectId, ref: 'Student' },
    schedule: {
        start: Number,
        end: Number,
        weekday: Number,
        week: [Number]
    },
    semester: {
        start: Number,
        name: String
    },
    createdAt: Number,
    updatedAt: Number
}, {
    timestamps: {
        currentTime: () => Date.now()
    }
});

const model = mongoose.model(
    "Course",
    schema
);

export const Course = model;

File: prj3_app/lib/models/course.model.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:prj3_app/helpers/converter.dart';

part "course.model.freezed.dart";
part "course.model.g.dart";

@freezed
class Course with _$Course {
  @JsonSerializable(explicitToJson: true)
  const factory Course({
    @JsonKey(name: "_id") String? id,
    @JsonKey(name: "id") required String courseId,
    required String name,
    required double midFactor,
    required int courseFactor,
    required Schedule schedule,
    required Semester semester,
    @DateTimeConverter() DateTime? createdAt,
    @DateTimeConverter() DateTime? updatedAt,
  }) = _Course;

  factory Course.fromJson(Map<String, dynamic> json) => _$CourseFromJson(json);
}

@freezed
class Schedule with _$Schedule {
  @JsonSerializable(explicitToJson: true)
  const factory Schedule(
      {@JsonKey(name: "_id") required String id,
      required int start,
      required int end,
      required int weekday,
      required List<int> week}) = _Schedule;

  factory Schedule.fromJson(Map<String, dynamic> json) => _$ScheduleFromJson(json);
}

@freezed
class Semester with _$Semester {
  @JsonSerializable(explicitToJson: true)
  const factory Semester({required int start, required String name}) = _Semester;

  factory Semester.fromJson(Map<String, dynamic> json) => _$SemesterFromJson(json);
}
